src/eepromfs.c

```
@@ -0,0 +1,933 @@
+    }
+
+    /* File does not exist */
+    return EEPFS_ENOFILE;
+}
+
+/**
+ * @brief Resolves a file handle to a file descriptor.
+ * 
+ * File descriptors can be used for reading or writing.
+ * 
+ * Getting a file repeatedly will always return the same pointer.
+ * This means that the cursor is shared across all usage of the file.
+ * Be sure to reset the cursor when you are done with it!
+ * 
+ * Negative handles indicate an upstream error occurred.
+ * The zero value will never be used as a file handle, even
+ * though it does map to an entry in the `eepfs_files` table.
+ * The signature entry is not intended to be used as a file.
+ * 
+ * @param[in] handle
+ *            The file handle of an entry in the `eepfs_files` table
+ * 
+ * @return A pointer to the file descriptor or NULL if the handle is invalid
+ */
+static eepfs_file_t * const eepfs_get_file(const int handle)
+{
+    /* Check if the handle appears to be valid */
+    if ( handle > 0 && handle < eepfs_files_count )
+    {
+        return &eepfs_files[handle];
+    }
+
+    return NULL;
+}
+
+/**
+ * @brief Reads a specific amount of data from a file starting from the cursor.
+ *
+ * @param[in]  path
+ *             Path of file in EEPROM filesystem to read from
+ * @param[out] dest
+ *             Buffer to read into
+ * @param[in]  len
+ *             Number of bytes to read
+ *
+ * @return The actual number of bytes read or a negative value on failure.
+ */
+static int eepfs_cursor_read(eepfs_file_t * const file, void * const dest, const size_t len)
+{
+    if ( file == NULL )
+    {
+        return EEPFS_EBADHANDLE;
+    }
+
+    if ( dest == NULL ) 
+    {
+        return EEPFS_EBADINPUT;
+    }
+
+    size_t cursor = file->cursor;
+    size_t bytes_to_read = len;
+
+    /* Bounds check to make sure it doesn't read past the end */
+    if ( cursor + bytes_to_read >= file->max_cursor )
+    {
+        /* It will, let's shorten it */
+        bytes_to_read = file->max_cursor - cursor;
```

Maybe we should return an error here as well. The buffer will not be filled by len bytes and that may cause confusion.

---

I borrowed this pattern from `dragonfs.c`: https://github.com/DragonMinded/libdragon/blob/87bcff52696abe21a1a322a895677d26ce764e46/src/dragonfs.c#L949-L957

---

It is a really old code and we don't have to duplicate that, but for the sake of consistency I think this is ok for now.